File: 2023/domaci/43-kontrola-pravopisu/volfmatous/main.py
if __name__ == "__main__":
    dictionary_words = set()

    d = int(input())
    for _ in range(d):
        dictionary_words.add(input())

    input()

    sentence_starts = True
    previous_word_was_number = False

    l = int(input())
    for _ in range(l):
        line = input()

        print(line)

        if len(line) < 1:
            continue

        error_indexes = []
        error_lengths = []
        error_messages = []

        current_index = 0
        duplicate_space_flag = False

        words = line.split(" ")

        for w in words:
            word = w

            if len(word) < 1:
                if duplicate_space_flag:
                    error_lengths[-1] += 1
                else:
                    error_indexes.append(max(0, current_index - 1))
                    error_lengths.append(2)
                    error_messages.append("duplicate space")
                    duplicate_space_flag = True

                current_index += 1
                continue

            duplicate_space_flag = False

            ends_with_dot = False
            if word.endswith("."):
                word = word[:-1]
                ends_with_dot = True

            is_number = True
            for c in word:
                if not ("0" <= c <= "9"):
                    is_number = False
                    break

            word_capitalized = word.capitalize()
            word_lowercase = word.lower()

            if is_number:
                if len(word) > 3:
                    error_indexes.append(current_index)
                    error_lengths.append(len(word))
                    error_messages.append("incorrect number formatting")
            elif word_capitalized not in dictionary_words and word_lowercase not in dictionary_words:
                error_indexes.append(current_index)
                error_lengths.append(len(word))
                error_messages.append("unknown word")
            elif word[0].isupper() and not sentence_starts and not previous_word_was_number and word not in dictionary_words:
                error_indexes.append(current_index)
                error_lengths.append(len(word))
                error_messages.append("incorrect leading uppercase")
            elif word[0].islower() and (sentence_starts or word_lowercase not in dictionary_words):
                error_indexes.append(current_index)
                error_lengths.append(len(word))
                error_messages.append("incorrect leading lowercase")

            current_index += len(w) + 1
            sentence_starts = ends_with_dot and not is_number
            previous_word_was_number = is_number

        if not error_indexes:
            continue

        current_index = 0
        for j in range(len(error_indexes)):
            for _ in range(error_indexes[j] - current_index):
                print(" ", end="")
            for _ in range(error_lengths[j]):
                print("~", end="")
            current_index = error_indexes[j] + error_lengths[j]

        print()

        current_index = 0
        for j in range(len(error_indexes)):
            for _ in range(error_indexes[j] - current_index):
                print(" ", end="")
            print("^", end="")
            current_index = error_indexes[j] + 1

        print()

        for line_index in range(len(error_indexes) * 2):
            current_index = 0
            for j in range(len(error_indexes) - line_index // 2):
                for _ in range(error_indexes[j] - current_index):
                    print(" ", end="")
                if j == len(error_indexes) - (line_index + 1) // 2:
                    print("-----", error_messages[j], end="")
                else:
                    print("|", end="")
                current_index = error_indexes[j] + 1
            print()

        print()

File: 2023/domaci/43-kontrola-pravopisu/volfmatous/main.cs
using System;
using System.Collections.Generic;

internal class Program
{
    static void Main(string[] args)
    {
        HashSet<string> dictionaryWords = new HashSet<string>();

        int d = int.Parse(Console.ReadLine());
        for (int i = 0; i < d; i++)
        {
            dictionaryWords.Add(Console.ReadLine());
        }

        Console.ReadLine();

        bool sentenceStarts = true;
        bool previousWordWasNumber = false;

        int l = int.Parse(Console.ReadLine());
        for (int i = 0; i < l; i++)
        {
            string line = Console.ReadLine();

            Console.WriteLine(line);

            if (line.Length < 1)
            {
                continue;
            }

            List<int> errorIndexes = new List<int>();
            List<int> errorLengths = new List<int>();
            List<string> errorMessages = new List<string>();

            int currentIndex = 0;
            bool duplicateSpaceFlag = false;

            string[] words = line.Split(' ');

            foreach (string w in words)
            {
                string word = w;

                if (word.Length < 1)
                {
                    if (duplicateSpaceFlag)
                    {
                        errorLengths[errorLengths.Count - 1]++;
                    }
                    else
                    {
                        errorIndexes.Add(Math.Max(0, currentIndex - 1));
                        errorLengths.Add(2);
                        errorMessages.Add("duplicate space");
                        duplicateSpaceFlag = true;
                    }

                    currentIndex++;
                    continue;
                }

                duplicateSpaceFlag = false;

                bool endsWithDot = false;
                if (word.EndsWith('.'))
                {
                    word = word.Substring(0, word.Length - 1);
                    endsWithDot = true;
                }

                bool isNumber = true;
                foreach (char c in word)
                {
                    if (!char.IsDigit(c))
                    {
                        isNumber = false;
                        break;
                    }
                }

                string wordCapitalized = char.ToUpper(word[0]) + word.Substring(1);
                string wordLowercase = word.ToLower();

                if (isNumber)
                {
                    if (word.Length > 3)
                    {
                        errorIndexes.Add(currentIndex);
                        errorLengths.Add(word.Length);
                        errorMessages.Add("incorrect number formatting");
                    }
                }
                else if (!dictionaryWords.Contains(wordCapitalized) && !dictionaryWords.Contains(wordLowercase))
                {
                    errorIndexes.Add(currentIndex);
                    errorLengths.Add(word.Length);
                    errorMessages.Add("unknown word");
                }
                else if (char.IsUpper(word[0]) && !sentenceStarts && !previousWordWasNumber && !dictionaryWords.Contains(word))
                {
                    errorIndexes.Add(currentIndex);
                    errorLengths.Add(word.Length);
                    errorMessages.Add("incorrect leading uppercase");
                }
                else if (char.IsLower(word[0]) && (sentenceStarts || !dictionaryWords.Contains(wordLowercase)))
                {
                    errorIndexes.Add(currentIndex);
                    errorLengths.Add(word.Length);
                    errorMessages.Add("incorrect leading lowercase");
                }

                currentIndex += w.Length + 1;
                sentenceStarts = endsWithDot && !isNumber;
                previousWordWasNumber = isNumber;
            }

            if (errorIndexes.Count < 1)
            {
                continue;
            }

            currentIndex = 0;
            for (int j = 0; j < errorIndexes.Count; j++)
            {
                for (int k = 0; k < errorIndexes[j] - currentIndex; k++)
                {
                    Console.Write(' ');
                }

                for (int k = 0; k < errorLengths[j]; k++)
                {
                    Console.Write('~');
                }

                currentIndex = errorIndexes[j] + errorLengths[j];
            }

            Console.WriteLine();

            currentIndex = 0;
            for (int j = 0; j < errorIndexes.Count; j++)
            {
                for (int k = 0; k < errorIndexes[j] - currentIndex; k++)
                {
                    Console.Write(' ');
                }

                Console.Write('^');

                currentIndex = errorIndexes[j] + 1;
            }

            Console.WriteLine();

            for (int lineIndex = 0; lineIndex < errorIndexes.Count * 2; lineIndex++)
            {
                currentIndex = 0;

                for (int j = 0; j < errorIndexes.Count - lineIndex / 2; j++)
                {
                    for (int k = 0; k < errorIndexes[j] - currentIndex; k++)
                    {
                        Console.Write(' ');
                    }

                    if (j == errorIndexes.Count - (lineIndex + 1) / 2)
                    {
                        Console.Write("----- " + errorMessages[j]);
                    }
                    else
                    {
                        Console.Write('|');
                    }

                    currentIndex = errorIndexes[j] + 1;
                }

                Console.WriteLine();
            }

            Console.WriteLine();
        }
    }
}
